Python/Module2_EssentialsOfPython/Errors_and_Debugging.md

```diff
@@ -0,0 +1,112 @@
+    text_representation:
+      extension: .md
+      format_name: markdown
+      format_version: '1.2'
+      jupytext_version: 1.3.0rc1
+  kernelspec:
+    display_name: Python 3
+    language: python
+    name: python3
+---
+
+<!-- #raw raw_mimetype="text/restructuredtext" -->
+.. meta::
+   :description: Topic: Basics of Python Objects, Difficulty: Easy, Category: Section
+   :keywords: exceptions, errors, bugs, debug, fix
+<!-- #endraw -->
+
+<!-- #region -->
+# Errors
+
+Unfortunately, not all code is perfect. As you write more complicated programs, you will doubtless run into some errors that you will need to figure out how to fix. Although debugging skills will come with practice, it can be helpful to understand errors, what causes them, and how to fix them. The different types of errors can be broken down into three categories: syntax errors, exceptions, and logic errors.
+
+
+<!-- #endregion -->
+
+<!-- #region -->
+## Syntax Errors
+
+Syntax errors are by far the easiest errors to debug, and they will be the most common errors while you are learning to program. Syntax errors are raised when the Python interpreter is unable to understand your code due to a typo or some misplaced syntax. When a syntax error occurs, the interpreter will stop running. It will also display the `SyntaxError` code, and will often give a line number or explanation.
+
+Most syntax errors are benign and easily detectable, such as mispelling a name or forgetting a piece of punctuation. Occasionally you will encounter a tricky syntax error where you cannot find your mistake in the line. When this happens, remember to look out for punctuation mistakes, especially in parentheses. Misplaced parentheses can be difficult to detect, but are relatively simple to fix once detected.
+
+Many modern IDEs will automatically detect syntax errors, and will sometimes automatically place punctuation and indentations. As such, syntax errors are becoming more easily avoidable, but it is still important to know how to spot and fix them. 
+
+<!-- #endregion -->
+
+<!-- #region -->
+## Exceptions
+An exception occurs when a piece of code is valid and understood by the interpreter, but for some reason the interpreter is unable to execute the action. Exceptions can occur when a function is passed an incompatible type of input, when the program is asked to something it cannot, and in many other circumstances. Python has many types of built-in exceptions that can be raised, but here are several that you are likely to encounter and their causes:
+
+| Exception  | Causes |
+| ------------- |:-------------:|
+| `IndexError` | Attempting to index or slice into a list with an index outside of the list's range |   
+| `KeyError` | Attempting to access a dictionary value with a key that is not in the dictionary |   
+| `NameError` | Referencing a variable or function that is not yet defined |   
+| `TypeError` | Passing an object of an invalid type into a function or operation |   
```

Comment:
My only thought would be that this could pop up when libraries are used - passing an invalid type into a lot of library functions will raise a TypeError. I know numpy uses a lot of TypeErrors, and I'm sure other libraries do as well -